research/flamby/fed_heart_disease/fedper/client.py

```diff
@@ -0,0 +1,141 @@
+import argparse
+import os
+from logging import INFO
+from pathlib import Path
+from typing import Optional, Sequence, Tuple
+
+import flwr as fl
+import torch
+import torch.nn as nn
+from flamby.datasets.fed_heart_disease import BATCH_SIZE, LR, NUM_CLIENTS, BaselineLoss
+from flwr.common.logger import log
+from flwr.common.typing import Config
+from torch.nn.modules.loss import _Loss
+from torch.optim import Optimizer
+from torch.utils.data import DataLoader
+
+from fl4health.checkpointing.checkpointer import BestMetricTorchCheckpointer, TorchCheckpointer
+from fl4health.clients.moon_client import MoonClient
+from fl4health.utils.losses import LossMeterType
+from fl4health.utils.metrics import Accuracy, Metric
+from fl4health.utils.random import set_all_random_seeds
+from research.flamby.fed_heart_disease.fedper.fedper_model import FedHeartDiseaseFedPerModel
+from research.flamby.flamby_data_utils import construct_fed_heard_disease_train_val_datasets
+
+
+class FedHeartDiseaseFedPerClient(MoonClient):
+    def __init__(
+        self,
+        data_path: Path,
+        metrics: Sequence[Metric],
+        device: torch.device,
+        client_number: int,
+        learning_rate: float,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        checkpointer: Optional[TorchCheckpointer] = None,
+    ) -> None:
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            checkpointer=checkpointer,
+        )
+        self.client_number = client_number
+        self.learning_rate: float = learning_rate
+
+        assert 0 <= client_number < NUM_CLIENTS
+        log(INFO, f"Client Name: {self.client_name}, Client Number: {self.client_number}")
+
+    def get_data_loaders(self, config: Config) -> Tuple[DataLoader, DataLoader]:
+        train_dataset, validation_dataset = construct_fed_heard_disease_train_val_datasets(
+            self.client_number, str(self.data_path)
+        )
+        train_loader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True)
+        val_loader = DataLoader(validation_dataset, batch_size=BATCH_SIZE, shuffle=False)
+        return train_loader, val_loader
+
+    def get_model(self, config: Config) -> nn.Module:
+        model: nn.Module = FedHeartDiseaseFedPerModel().to(self.device)
+        return model
+
+    def get_optimizer(self, config: Config) -> Optimizer:
+        return torch.optim.AdamW(self.model.parameters(), lr=self.learning_rate)
+
+    def get_criterion(self, config: Config) -> _Loss:
+        return BaselineLoss()
+
```

Comment:
I may be missing something, so please correct me if I'm wrong, but is there a place where you define the partial weight exchanger for FedPer (similar to `examples/fedper_example/client.py`). Without that definition, I think we default to a full weight exchange approach, which isn't what we want. That is, we want only the base module to be exchanged and the classifier layer to stay local. I may be missing where that happens though 🙂

Reply:
That is true, I think I have missed that we should also implement `get_parameter_exchanger` for fedper. I will update and rerun the experiments.